modules/schema/src/main/java/org/apache/ignite/internal/schema/mapping/ColumnMapperImpl.java

```diff
@@ -17,34 +17,56 @@
 
 package org.apache.ignite.internal.schema.mapping;
 
+import org.apache.ignite.internal.schema.Column;
+import org.apache.ignite.internal.schema.SchemaDescriptor;
+import org.jetbrains.annotations.NotNull;
+
 /**
  * Column mapper implementation.
  */
 class ColumnMapperImpl implements ColumnMapper, ColumnMapperBuilder {
     /** Mapping. */
     private final int[] mapping;
 
+    /** Mapped columns. */
+    private final Column[] cols;
+
     /**
-     * @param cols Number of columns.
+     * @param schema Schema descriptor.
      */
-    ColumnMapperImpl(int cols) {
-        mapping = new int[cols];
+    ColumnMapperImpl(SchemaDescriptor schema) {
+        mapping = new int[schema.length()];
+        cols = new Column[schema.length()];
 
-        for (int i = 0; i < mapping.length; i++)
+        for (int i = 0; i < mapping.length; i++) {
             mapping[i] = i;
+            cols[i] = schema.column(i);
+        }
     }
 
     /** {@inheritDoc} */
-    @Override public void add(int from, int to) {
+    @Override public ColumnMapperImpl add(@NotNull Column col) {
+        add(col.schemaIndex(), -1, col);
```

Comment:
this invocation violates contract from javadoc: `col` is a target column descriptor, but if there is no such column in target (to==-1), then there should not be any descriptor

Reply:
This means a new column was added and the old schema has no column the given one is mapped to.
So, we have to save the poison index (-1) and the current descriptor to be able to see a correct default value during further updates.